File: lessons/04-services/python/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial #To be able to pass extra data to the callback

class AddTwoIntsClient(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.client_=self.create_client(srv_type=AddTwoInts,#service type
                                        srv_name="add_two_ints")#service name
        
    def call_add_two_ints(self,a:int,b:int):
        while not self.client_.wait_for_service(timeout_sec=1.0):#Wait for server to start for 1sec, if not started, display the following warning
            self.get_logger().warn("Waiting for server.....")

        request=AddTwoInts.Request()
        request.a=a
        request.b=b

        future=self.client_.call_async(request)#using asynchronous call as normal call might cause deadlock
        future.add_done_callback(partial(self.callback_call_add_two_ints,request=request))#callback until response arrives(can't spin inside a spinning node)
    
    def callback_call_add_two_ints(self, future,request):
        response=future.result()
        self.get_logger().info(f"{request.a}+{request.b}={response.sum}")


def main(args=None):
    rclpy.init(args=args)
    node=AddTwoIntsClient()
    node.call_add_two_ints(2,3)
    rclpy.spin(node)
    rclpy.shutdown()

File: lessons/04-services/cpp/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
using namespace std::chrono_literals;
using namespace std::placeholders;

class AddTwoIntsClient : public rclcpp::Node
{
    public:
    AddTwoIntsClient(): Node("add_two_ints_client") {
        client_=this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");//create client with service name and service type
    }
    
    void callAddTwoInts(int a, int b) { //function to call the service
        while (!client_->wait_for_service(1s)) {//keep checking every second and display the following warning until the server is available
            RCLCPP_WARN(this->get_logger(),"Waiting for the server");
        }
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a=a;
        request->b=b;

        client_->async_send_request(
            request, std::bind(&AddTwoIntsClient::callbackAddTwoInts, this, _1)//placeholder for SharedFuture object
        );


    }

    private:
    rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_;//declare client as a shared pointer
    void callbackAddTwoInts(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future) {
        auto response=future.get();
        RCLCPP_INFO(this->get_logger(),"Sum: %d",(int)response->sum);
    }
};

int main(int argc, char **argv){
    rclcpp::init(argc,argv);
    auto node=std::make_shared<AddTwoIntsClient>();
    node->callAddTwoInts(7,6);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}